File: Lernwerk/Services/Research/Wikipedia.swift
import Foundation
#if canImport(FoundationNetworking)
import FoundationNetworking
#endif

/// A Wikipedia article or the start of it: title, link and plain text.
struct WikiArticle: Equatable {
    var title: String
    var url: String
    var text: String
}

/// Reads the German Wikipedia through its public API; no key needed. Mirrors the Android app.
struct WikipediaClient {
    /// One GET: URL and headers in, body and status code out. Replaced in tests.
    typealias Fetch = (URL, [String: String]) async throws -> (Data, Int)

    static let userAgent = "SchulPip/1.0 (https://github.com/Maxifrz/schul-pip; Lern-App für Schüler)"

    var language = "de"
    var fetch: Fetch = WikipediaClient.fetchWithURLSession

    private var api: String { "https://\(language).wikipedia.org/w/api.php" }

    /// The best matches for `query` with the introduction of each article, best first.
    func search(_ query: String, limit: Int = 2) async throws -> [WikiArticle] {
        let query = query.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !query.isEmpty else { return [] }
        let response = try await get([
            ("action", "query"),
            ("generator", "search"),
            ("gsrsearch", query),
            ("gsrlimit", "\(limit)"),
            ("gsrnamespace", "0"),
            ("prop", "extracts|info"),
            ("inprop", "url"),
            ("exintro", "1"),
            ("explaintext", "1"),
            ("exlimit", "\(limit)"),
            ("redirects", "1"),
        ])
        return pages(response).sorted { $0.rank < $1.rank }.map(\.article).filter { !$0.text.isEmpty }
    }

    /// The article's text up to `maxChars`, cut at the end of a sentence.
    func article(_ title: String, maxChars: Int = 3500) async throws -> WikiArticle? {
        let response = try await get([
            ("action", "query"),
            ("titles", title),
            ("prop", "extracts|info"),
            ("inprop", "url"),
            ("explaintext", "1"),
            ("exsectionformat", "plain"),
            ("redirects", "1"),
        ])
        guard var article = pages(response).first?.article else { return nil }
        article.text = WikiText.shorten(article.text, maxChars: maxChars)
        return article.text.isEmpty ? nil : article
    }

    /// The introduction of the article that best matches `query`, for a quick explanation.
    func summary(_ query: String, maxChars: Int = 1200) async throws -> WikiArticle? {
        guard var article = try await search(query, limit: 1).first else { return nil }
        article.text = WikiText.shorten(article.text, maxChars: maxChars)
        return article
    }

    private struct Response: Decodable {
        struct Query: Decodable {
            var pages: [Page]?
        }

        struct Page: Decodable {
            var title: String?
            var extract: String?
            var fullurl: String?
            var index: Int?
            var missing: Bool?
            var invalid: Bool?
        }

        var query: Query?
    }

    private func get(_ parameters: [(String, String)]) async throws -> Response {
        let query = (parameters + [("format", "json"), ("formatversion", "2")])
            .map { "\($0.0)=\(Self.encode($0.1))" }
            .joined(separator: "&")
        guard let url = URL(string: "\(api)?\(query)") else { throw LLMError.invalidResponse }
        let (data, status) = try await fetch(url, ["User-Agent": Self.userAgent, "Accept": "application/json"])
        guard (200..<300).contains(status) else { throw LLMError.http(status: status, message: "Wikipedia") }
        guard let response = try? JSONDecoder().decode(Response.self, from: data) else { throw LLMError.invalidResponse }
        return response
    }

    /// Articles with their search rank; missing pages are left out.
    private func pages(_ response: Response) -> [(article: WikiArticle, rank: Int)] {
        (response.query?.pages ?? []).compactMap { page in
            guard page.missing != true, page.invalid != true, let title = page.title else { return nil }
            let url = page.fullurl ?? "https://\(language).wikipedia.org/wiki/\(title.replacingOccurrences(of: " ", with: "_"))"
            let text = (page.extract ?? "").trimmingCharacters(in: .whitespacesAndNewlines)
            return (WikiArticle(title: title, url: url, text: text), page.index ?? Int.max)
        }
    }

    /// Percent-encodes everything but the unreserved characters of RFC 3986.
    static func encode(_ value: String) -> String {
        let allowed = CharacterSet(charactersIn: "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789-._~")
        return value.addingPercentEncoding(withAllowedCharacters: allowed) ?? value
    }

    static func fetchWithURLSession(_ url: URL, headers: [String: String]) async throws -> (Data, Int) {
        var request = URLRequest(url: url, timeoutInterval: 20)
        for (name, value) in headers {
            request.setValue(value, forHTTPHeaderField: name)
        }
        return try await URLSession.shared.llmData(for: request)
    }
}

enum WikiText {
    /// At most `maxChars` characters, ending after a full sentence where possible; blank lines are collapsed.
    static func shorten(_ text: String, maxChars: Int) -> String {
        let clean = text.replacingOccurrences(of: "\n{3,}", with: "\n\n", options: .regularExpression)
            .trimmingCharacters(in: .whitespacesAndNewlines)
        guard clean.count > maxChars else { return clean }
        let cut = String(clean.prefix(max(0, maxChars)))
        let ends = [cut.range(of: ". ", options: .backwards), cut.range(of: ".\n", options: .backwards)]
            .compactMap { $0 }
            .map { cut.distance(from: cut.startIndex, to: $0.lowerBound) }
        if let end = ends.max(), end >= maxChars / 3 {
            return String(cut.prefix(end + 1))
        }
        return cut + " …"
    }
}

/// An article the AI may use, with the id it cites it by.
struct WebSource: Equatable {
    var id: String
    var title: String
    var url: String
    var text: String
}

/// Collects Wikipedia articles for a presentation and turns them into prompt text and citations.
enum Research {
    static let maxQueries = 4
    static let maxCharacters = 14_000

    /// For every query the introductions of the two best matches and a longer excerpt of the best one; articles that
    /// are already there are skipped. A query that fails is skipped too, the talk can still be built without it.
    static func gather(_ wikipedia: WikipediaClient, queries: [String], existing: [WebSource] = []) async throws -> [WebSource] {
        var sources = existing
        var characters = sources.reduce(0) { $0 + $1.text.count }
        var seen = Set<String>()
        let cleaned = queries.map { $0.trimmingCharacters(in: .whitespacesAndNewlines) }
            .filter { !$0.isEmpty && seen.insert($0).inserted }
            .prefix(maxQueries)
        for query in cleaned {
            try Task.checkCancellation()
            guard let found = try? await wikipedia.search(query, limit: 2) else { continue }
            for (index, hit) in found.enumerated() {
                guard !sources.contains(where: { $0.title == hit.title }), characters < maxCharacters else { continue }
                var text = WikiText.shorten(hit.text, maxChars: 1200)
                if index == 0, let article = try? await wikipedia.article(hit.title) {
                    text = article.text
                }
                let budgeted = WikiText.shorten(text, maxChars: maxCharacters - characters)
                guard !budgeted.isEmpty else { continue }
                sources.append(WebSource(id: "W\(sources.count + 1)", title: hit.title, url: hit.url, text: budgeted))
                characters += budgeted.count
            }
        }
        return sources
    }

    /// The articles as a block for the prompt, each with its id.
    static func prompt(_ sources: [WebSource]) -> String {
        var lines = ["<research source=\"German Wikipedia\">"]
        for source in sources {
            lines.append("<article id=\"\(source.id)\" title=\"\(source.title)\" url=\"\(source.url)\">")
            lines.append(source.text)
            lines.append("</article>")
        }
        lines.append("</research>")
        return lines.joined(separator: "\n")
    }

    /// How a school talk cites an article: title, Wikipedia, link and the day it was read.
    static func citation(_ source: WebSource, date: Date = Date()) -> String {
        let formatter = DateFormatter()
        formatter.locale = Locale(identifier: "de_DE")
        formatter.dateFormat = "dd.MM.yyyy"
        let link = source.url.hasPrefix("https://") ? String(source.url.dropFirst("https://".count)) : source.url
        return "„\(source.title)“, Wikipedia, \(link) (abgerufen am \(formatter.string(from: date)))"
    }

    /// A link to YouTube's search for `query`; opens the app or the website, needs no key.
    static func youtubeSearchURL(_ query: String) -> URL? {
        let terms = query.trimmingCharacters(in: .whitespacesAndNewlines)
            .split(whereSeparator: \.isWhitespace)
            .map { WikipediaClient.encode(String($0)) }
            .joined(separator: "+")
        return URL(string: "https://www.youtube.com/results?search_query=\(terms)")
    }
}

File: android/app/src/main/java/de/maxifrz/lernwerk/research/Wikipedia.kt
package de.maxifrz.lernwerk.research

import de.maxifrz.lernwerk.llm.HttpTransport
import de.maxifrz.lernwerk.llm.LlmError
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.contentOrNull
import kotlinx.serialization.json.intOrNull
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import java.net.URLEncoder
import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale

/** A Wikipedia article or the start of it: title, link and plain text. */
data class WikiArticle(val title: String, val url: String, val text: String)

/** Reads the German Wikipedia through its public API; no key needed. */
class WikipediaClient(private val transport: HttpTransport, private val language: String = "de") {
    private val api = "https://$language.wikipedia.org/w/api.php"

    /** The best matches for [query] with the introduction of each article, best first. */
    suspend fun search(query: String, limit: Int = 2): List<WikiArticle> {
        if (query.isBlank()) return emptyList()
        val root = get(
            "action" to "query",
            "generator" to "search",
            "gsrsearch" to query.trim(),
            "gsrlimit" to limit.toString(),
            "gsrnamespace" to "0",
            "prop" to "extracts|info",
            "inprop" to "url",
            "exintro" to "1",
            "explaintext" to "1",
            "exlimit" to limit.toString(),
            "redirects" to "1",
        )
        return pages(root).sortedBy { it.second }.map { it.first }.filter { it.text.isNotBlank() }
    }

    /** The article's text up to [maxChars], cut at the end of a sentence. */
    suspend fun article(title: String, maxChars: Int = 3500): WikiArticle? {
        val root = get(
            "action" to "query",
            "titles" to title,
            "prop" to "extracts|info",
            "inprop" to "url",
            "explaintext" to "1",
            "exsectionformat" to "plain",
            "redirects" to "1",
        )
        val article = pages(root).firstOrNull()?.first ?: return null
        return article.copy(text = WikiText.shorten(article.text, maxChars)).takeIf { it.text.isNotBlank() }
    }

    /** The introduction of the article that best matches [query], for a quick explanation. */
    suspend fun summary(query: String, maxChars: Int = 1200): WikiArticle? =
        search(query, limit = 1).firstOrNull()?.let { it.copy(text = WikiText.shorten(it.text, maxChars)) }

    private suspend fun get(vararg parameters: Pair<String, String>): JsonObject {
        val query = (parameters.toList() + listOf("format" to "json", "formatversion" to "2"))
            .joinToString("&") { (name, value) -> "$name=${URLEncoder.encode(value, "UTF-8")}" }
        val result = transport.get("$api?$query", mapOf("User-Agent" to USER_AGENT, "Accept" to "application/json"), TIMEOUT_SECONDS)
        if (result.status !in 200..299) throw LlmError.Http(result.status, "Wikipedia")
        return runCatching { Json.parseToJsonElement(result.body).jsonObject }.getOrElse { throw LlmError.InvalidResponse }
    }

    /** Articles with their search rank; missing pages are left out. */
    private fun pages(root: JsonObject): List<Pair<WikiArticle, Int>> {
        val pages = (root["query"] as? JsonObject)?.get("pages") as? JsonArray ?: return emptyList()
        return pages.mapNotNull { element ->
            val page = element as? JsonObject ?: return@mapNotNull null
            if (page["missing"] != null || page["invalid"] != null) return@mapNotNull null
            val title = page["title"]?.jsonPrimitive?.contentOrNull ?: return@mapNotNull null
            val url = page["fullurl"]?.jsonPrimitive?.contentOrNull ?: "https://$language.wikipedia.org/wiki/${title.replace(' ', '_')}"
            val text = page["extract"]?.jsonPrimitive?.contentOrNull?.trim() ?: ""
            WikiArticle(title, url, text) to (page["index"]?.jsonPrimitive?.intOrNull ?: Int.MAX_VALUE)
        }
    }

    companion object {
        const val USER_AGENT = "SchulPip/1.0 (https://github.com/Maxifrz/schul-pip; Lern-App für Schüler)"
        const val TIMEOUT_SECONDS = 20L
    }
}

object WikiText {
    /** At most [maxChars] characters, ending after a full sentence where possible; blank lines are collapsed. */
    fun shorten(text: String, maxChars: Int): String {
        val clean = text.replace(Regex("\n{3,}"), "\n\n").trim()
        if (clean.length <= maxChars) return clean
        val cut = clean.take(maxChars)
        val end = maxOf(cut.lastIndexOf(". "), cut.lastIndexOf(".\n"))
        return if (end >= maxChars / 3) cut.take(end + 1) else "$cut …"
    }
}

/** An article the AI may use, with the id it cites it by. */
data class WebSource(val id: String, val title: String, val url: String, val text: String)

/** Collects Wikipedia articles for a presentation and turns them into prompt text and citations. */
object Research {
    const val MAX_QUERIES = 4
    const val MAX_CHARACTERS = 14_000

    /**
     * For every query the introductions of the two best matches and a longer excerpt of the best one; articles that
     * are already there are skipped. A query that fails is skipped too, the talk can still be built without it.
     */
    suspend fun gather(
        wikipedia: WikipediaClient,
        queries: List<String>,
        existing: List<WebSource> = emptyList(),
    ): List<WebSource> {
        val sources = existing.toMutableList()
        var characters = sources.sumOf { it.text.length }
        for (query in queries.map { it.trim() }.filter { it.isNotEmpty() }.distinct().take(MAX_QUERIES)) {
            val found = try {
                wikipedia.search(query, limit = 2)
            } catch (error: kotlinx.coroutines.CancellationException) {
                throw error
            } catch (error: Exception) {
                continue
            }
            found.forEachIndexed { index, hit ->
                if (sources.any { it.title == hit.title } || characters >= MAX_CHARACTERS) return@forEachIndexed
                val text = if (index == 0) {
                    runCatching { wikipedia.article(hit.title)?.text }.getOrNull() ?: hit.text
                } else {
                    WikiText.shorten(hit.text, 1200)
                }
                val budgeted = WikiText.shorten(text, MAX_CHARACTERS - characters)
                if (budgeted.isBlank()) return@forEachIndexed
                sources += WebSource("W${sources.size + 1}", hit.title, hit.url, budgeted)
                characters += budgeted.length
            }
        }
        return sources
    }

    /** The articles as a block for the prompt, each with its id. */
    fun prompt(sources: List<WebSource>): String = buildString {
        appendLine("<research source=\"German Wikipedia\">")
        sources.forEach { source ->
            appendLine("<article id=\"${source.id}\" title=\"${source.title}\" url=\"${source.url}\">")
            appendLine(source.text)
            appendLine("</article>")
        }
        append("</research>")
    }

    /** How a school talk cites an article: title, Wikipedia, link and the day it was read. */
    fun citation(source: WebSource, date: Date = Date()): String =
        "„${source.title}“, Wikipedia, ${source.url.removePrefix("https://")} (abgerufen am ${SimpleDateFormat("dd.MM.yyyy", Locale.GERMANY).format(date)})"

    /** A link to YouTube's search for [query]; opens the app or the website, needs no key. */
    fun youtubeSearchUrl(query: String): String =
        "https://www.youtube.com/results?search_query=${URLEncoder.encode(query.trim(), "UTF-8")}"
}
